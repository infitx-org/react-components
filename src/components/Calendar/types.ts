export type PossibleDate = Date | undefined;

export type PossibleDay = string | undefined;

export type Matrix = PossibleDay[][];

export type DisabledDays = (day: Date) => boolean;

export type DateRange = [PossibleDate, PossibleDate];
