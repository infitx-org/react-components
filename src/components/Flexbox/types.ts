export enum TopBottom {
  "top" = "flex-start",
  "bottom" = "flex-end",
}
export enum LeftRight {
  "left" = "flex-start",
  "right" = "flex-end",
}

export enum Mappers {
  "top" = "flex-start",
  "bottom" = "flex-end",
  "left" = "flex-start",
  "right" = "flex-end",
}

export type AlignItems =
  | "flex-start"
  | "flex-end"
  | "center"
  | "baseline"
  | "stretch";

export type JustifyContent =
  | "flex-start"
  | "flex-end"
  | "center"
  | "space-between"
  | "space-around"
  | "space-evenly";

export type AlignWithMap<T extends string> = `${AlignItems}` | `${T}`;
export type JustifyWithMap<T extends string> = `${JustifyContent}` | `${T}`;
