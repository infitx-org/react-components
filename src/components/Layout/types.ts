export enum AlignMap {
  "top" = "flex-start",
  "bottom" = "flex-end",
}

export enum JustifyMap {
  "left" = "flex-start",
  "right" = "flex-end",
}

export type AlignItems =
  | "flex-start"
  | "flex-end"
  | "center"
  | "baseline"
  | "stretch";

export type JustifyContent =
  | "flex-start"
  | "flex-end"
  | "center"
  | "space-between"
  | "space-around"
  | "space-evenly";

export type AlignWithMap = `${AlignItems | keyof typeof AlignMap}`;
export type JustifyWithMap = `${JustifyContent | keyof typeof JustifyMap}`;
