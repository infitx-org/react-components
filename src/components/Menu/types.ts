export interface MenuSectionProps {
  label: string;
  hidden?: boolean;
  icon?: React.ReactElement<React.SVGProps<SVGSVGElement>>;
  fill?: string;
  size?: number;
  disabled?: boolean;
  children: MenuItemElement | MenuItemElement[];
}

interface BaseMenuItemProps {
  label: string;
  active?: boolean;
  disabled?: boolean;
  hidden?: boolean;
  back?: boolean;
  icon?: React.ReactElement<React.SVGProps<SVGSVGElement>>;
  fill?: string;
  size?: number;
  // prop partial is not used directly by the MenuItem but it is used by the Menu component
  partial?: boolean;
  children?: MenuItemElement | MenuElement[];
}

interface BaseMenuItemPropsWithTo extends BaseMenuItemProps {
  to: string;
  path?: string;
}
interface BaseMenuItemPropsWithPath extends BaseMenuItemProps {
  to?: string;
  path: string;
}

export type MenuItemProps = BaseMenuItemPropsWithTo | BaseMenuItemPropsWithPath;

export type MenuItemElement = React.ReactElement<MenuItemProps>;
export type MenuSectionElement = React.ReactElement<MenuSectionProps>;
export type MenuElement = MenuItemElement | MenuSectionElement;
