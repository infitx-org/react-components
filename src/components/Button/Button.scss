@use "sass:map";
@use "../../scss/base/colors.scss";
@use "../../scss/base/shadows.scss" as *;
@use "../../scss/base/transitions.scss" as *;
@use "../../scss/components/inputs.scss" as *;
@use "../../scss/components/any.scss" as *;
@use "../../scss/themes.scss" as *;
@use "../../scss/functions.scss" as *;

.rc-button {
  @include rc-component;

  outline: none;
  cursor: pointer;
  box-sizing: border-box;
  display: inline-block;
  flex: 0 0 auto;
  white-space: nowrap;
  user-select: none;
  text-decoration: none;
  transition: box-shadow 0.5s cubic-bezier(0.075, 0.82, 0.165, 1),
    color 0.2s cubic-bezier(0.075, 0.82, 0.165, 1), background 0.1s ease-out;
  border-radius: $input-border-radius;
  font-style: normal;
  font-weight: 400;
  -webkit-font-smoothing: antialiased;
  border: 1px solid;
  align-items: center;
  border-bottom-width: 2px;
  box-shadow: $size--closed transparent;

  &:active {
    color: colors.$alpha;
    fill: colors.$alpha;
    stroke: colors.$alpha;
  }

  &:hover,
  &:focus {
    border-bottom-width: 2px;
    border-bottom-color: rgba(0, 0, 0, 0.2);

    &:active {
      border-bottom-color: transparent;
    }
  }

  &:not(.rc-button--noFill):not(:disabled) {
    border-bottom-color: rgba(0, 0, 0, 0.2);
  }
}

.rc-button__content {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.rc-button__icon {
  display: flex;
  fill: inherit;
  stroke: inherit;
}

.rc-button__icon--left {
  margin-right: 5px;
  margin-left: 0;
}

.rc-button__icon--right {
  margin-right: 0;
  margin-left: 5px;
}

.rc-button--extra-small {
  padding: 4px 7px;
  font-size: inputfontsize($input-size--extra-small);
  height: $input-size--extra-small;
}

.rc-button--small {
  padding: 6px 10px;
  font-size: inputfontsize($input-size--small);
  height: $input-size--small;
}

.rc-button--medium {
  padding: 8px 13px;
  font-size: inputfontsize($input-size--medium);
  height: $input-size--medium;
}

.rc-button--large {
  padding: 10px 16px;
  font-size: inputfontsize($input-size--large);
  height: $input-size--large;
}

@each $kind in $kinds {
  .rc-button--#{$kind} {
    background-color: map.get($theme-kinds, $kind);
    border-color: map.get($theme-kinds, $kind);
    color: map.get($theme-kinds--opposite, $kind);
    fill: map.get($theme-kinds--opposite, $kind);
    stroke: map.get($theme-kinds--opposite, $kind);

    &.rc-button--noFill {
      background-color: transparent;
      color: map.get($theme-kinds--dark, $kind);
      fill: map.get($theme-kinds, $kind);
      stroke: map.get($theme-kinds, $kind);
    }

    &:active {
      color: colors.$alpha;
      fill: colors.$alpha;
      stroke: colors.$alpha;
    }

    &:active,
    &:focus {
      box-shadow: $size map.get($theme-kinds-shadows, $kind);
    }

    &:hover,
    &:focus {
      background-color: map.get($theme-kinds, $kind);
      border-color: map.get($theme-kinds, $kind);

      &.rc-button--noFill {
        color: map.get($theme-kinds--opposite, $kind);
        fill: map.get($theme-kinds--opposite, $kind);
        stroke: map.get($theme-kinds--opposite, $kind);
      }
    }

    &:hover:active {
      background: map.get($theme-kinds--dark, $kind);

      &.rc-button--noFill {
        color: colors.$alpha;
        fill: colors.$alpha;
        stroke: colors.$alpha;
      }
    }
  }
}

.rc-button--disabled {
  opacity: $input-opacity--disabled;
  pointer-events: none;
  cursor: auto;
}
