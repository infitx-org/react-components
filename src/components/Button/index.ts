export { default, BaseButton } from "./Button";
export type { ButtonProps } from "./Button";
