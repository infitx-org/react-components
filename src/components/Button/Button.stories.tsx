/* eslint no-console: "off" */
import { Size, Kind } from "types";
import Row from "components/Layout/Row";
import Button from "./Button";

export default {
  title: "Components/Button",
  component: Button,
};

const { log } = console;

const icon = (
  <svg
    xmlns="http://www.w3.org/2000/svg"
    width="40"
    height="40"
    viewBox="0 0 40 40"
  >
    <circle cx="20" cy="20" r="20" />
  </svg>
);

const Template = (args) => (
  <Button {...args} onClick={log} label="I am a button" />
);

export const Default = Template.bind({});
Default.args = {
  children: undefined,
  icon: undefined,
  label: undefined,
  className: undefined,
  id: undefined,
  kind: undefined,
  size: undefined,
  iconPosition: undefined,
  noFill: undefined,
  disabled: undefined,
  pending: undefined,
  style: undefined,
  onClick: undefined,
  onKeyDown: undefined,
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const Pending = Template.bind({});
Pending.args = {
  pending: true,
};

export const NoFill = Template.bind({});
NoFill.args = {
  noFill: true,
};

export const WithIcon = Template.bind({});
WithIcon.args = {
  icon,
};

export const WithIconOnTheRight = Template.bind({});
WithIconOnTheRight.args = {
  icon,
  iconPosition: "right",
};

const kinds = Object.values(Kind);
const sharedProps = (kind) => ({
  size: "medium",
  style: {
    margin: "2px",
  },
  onClick: log,
  label: kind,
  kind,
  key: kind,
});

export const Kinds = () => (
  <>
    <Row align="center space-between">
      {kinds.map((kind) => (
        <Button {...sharedProps(kind)} />
      ))}
    </Row>
    <Row align="center space-between">
      {kinds.map((kind) => (
        <Button {...sharedProps(kind)} noFill />
      ))}
    </Row>
  </>
);

export const KindsWithIcon = () => (
  <>
    <Row align="center space-between">
      {kinds.map((kind) => (
        <Button {...sharedProps(kind)} icon={icon} />
      ))}
    </Row>
    <Row align="center space-between">
      {kinds.map((kind) => (
        <Button {...sharedProps(kind)} noFill icon={icon} />
      ))}
    </Row>
  </>
);

export const KindsPeding = () => (
  <>
    <Row align="center space-between">
      {kinds.map((kind) => (
        <Button {...sharedProps(kind)} pending />
      ))}
    </Row>
    <Row align="center space-between">
      {kinds.map((kind) => (
        <Button {...sharedProps(kind)} noFill pending />
      ))}
    </Row>
  </>
);

export const Sizes = () => (
  <>
    <Row align="center space-between">
      {kinds.map((kind) => (
        <Button {...sharedProps(kind)} size="small" />
      ))}
    </Row>
    <Row align="center space-between">
      {kinds.map((kind) => (
        <Button {...sharedProps(kind)} size="medium" />
      ))}
    </Row>
    <Row align="center space-between">
      {kinds.map((kind) => (
        <Button {...sharedProps(kind)} size="large" />
      ))}
    </Row>
  </>
);
