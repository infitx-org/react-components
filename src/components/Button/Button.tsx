import React, { MouseEvent, KeyboardEvent } from "react";
import classnames from "classnames";
import "./Button.scss";

import { Size, Kind } from "../types";
import Spinner from "../Spinner";
import Icon from "../Icon";
import { getIconSizeByComponentSize } from "../shared";

export type ButtonProps = {
  children?: React.ReactNode;
  icon?: React.ReactElement<React.SVGProps<SVGSVGElement>>;
  label?: string;
  className?: string;
  id?: string;
  kind?: `${Kind}`;
  size?: `${Size}`;
  iconPosition?: "left" | "right";
  noFill?: boolean;
  disabled?: boolean;
  pending?: boolean;
  tooltip?: string;
  style?: React.CSSProperties;
  onClick: (e: MouseEvent<HTMLButtonElement>) => void;
  onKeyDown?: (e: KeyboardEvent<HTMLButtonElement>) => void;
};

function Button({
  children,
  id,
  className,
  style,
  label,
  noFill,
  disabled,
  pending,
  kind = Kind.Primary,
  size = Size.Large,
  icon,
  iconPosition = "left",
  onClick,
  onKeyDown,
}: ButtonProps): JSX.Element {
  const classNames = classnames([
    className,
    "mb-input",
    "input-button",
    `input-button--${kind}`,
    `input-button--${size}`,
    disabled && "mb-input--disabled input-button--disabled",
    pending && "mb-input--pending input-button--pending",
    noFill && "input-button--noFill",
  ]);

  let iconContent = null;
  if (icon || pending) {
    const numericSize = getIconSizeByComponentSize(size);
    let display;
    if (pending) {
      display = <Spinner color="inherit" size={numericSize} />;
    } else if (icon) {
      display = <Icon icon={icon} size={size} inheritFill />;
    }
    iconContent = (
      <div className={`input-button__icon input-button__icon--${iconPosition}`}>
        {display}
      </div>
    );
  }

  const button = (
    <button
      type="button"
      id={id}
      style={style}
      className={classNames}
      onKeyDown={onKeyDown}
      onClick={onClick}
      disabled={disabled}
    >
      <div className="input-button__content">
        {iconPosition === "left" && iconContent}
        {label || children}
        {iconPosition === "right" && iconContent}
      </div>
    </button>
  );

  return button;
}

export default React.memo(Button);
