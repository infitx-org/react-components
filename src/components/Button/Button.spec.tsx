import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";
import Button from "./Button";
import { Kind, Size } from "../types";

/* eslint-disable no-console */
const { log } = console;

const icon = (
  <svg width="55" height="55" viewBox="-6 -3 55 55">
    <circle cx="10" cy="10" r="20" />
  </svg>
);

describe("tests the buttton", () => {
  it("renders the label", () => {
    render(<Button label="Test-Button" onClick={log} />);
    expect(screen.getByText("Test-Button")).toBeDefined();
  });

  it("renders the children", () => {
    render(<Button onClick={log}>Test-Button</Button>);
    expect(screen.getByText("Test-Button")).toBeDefined();
  });

  it("renders the disabled state", () => {
    const { container } = render(
      <Button onClick={log} disabled>
        Test-Button
      </Button>
    );
    expect(container.querySelector("button")).toBeDisabled();
  });

  it("renders the spinner", () => {
    const { container } = render(<Button onClick={log} pending />);
    expect(container.querySelector(".el-spinner")).toBeTruthy();
  });

  it("renders the spinner before the icon", () => {
    const { container } = render(<Button icon={icon} onClick={log} pending />);
    expect(container.querySelector(".el-spinner")).toBeTruthy();
  });

  it("renders the icon", () => {
    const { container } = render(<Button icon={icon} onClick={log} />);
    expect(container.querySelector("svg")).toBeTruthy();
  });

  it("renders the id", () => {
    const { container } = render(<Button id="test" onClick={log} />);
    expect(container.querySelector("button#test")).toBeTruthy();
  });

  it("renders the className", () => {
    const { container } = render(<Button className="test" onClick={log} />);
    expect(container.querySelector("button.test")).toBeTruthy();
  });

  it("renders the style prop", () => {
    const { container } = render(
      <Button style={{ fill: "red" }} onClick={log} />
    );
    expect(container.querySelector("button")).toHaveAttribute(
      "style",
      "fill: red;"
    );
  });

  it("renders the default kind as primary kind", () => {
    const { container } = render(<Button onClick={log} />);
    expect(
      container.querySelector("button.input-button--primary")
    ).toBeTruthy();
  });

  it("renders all the kinds", () => {
    Object.values(Kind).forEach((kind) => {
      const { container } = render(<Button onClick={log} kind={kind} />);
      expect(
        container.querySelector(`button.input-button--${kind}`)
      ).toBeTruthy();
    });
  });

  it('renders with the "noFill" prop', () => {
    const { container } = render(<Button onClick={log} noFill />);
    expect(container.querySelector("button.input-button--noFill")).toBeTruthy();
  });

  it("renders the large, medium, small sizes", () => {
    Object.values(Size).forEach((size) => {
      const { container } = render(<Button onClick={log} size={size} />);
      expect(container.querySelector(`.input-button--${size}`)).toBeTruthy();
    });
  });

  it("triggers the onClick prop", () => {
    const mockEvent = jest.fn();
    const { container } = render(<Button onClick={mockEvent} />);
    userEvent.click(container.querySelector("button") as HTMLButtonElement);
    expect(mockEvent).toHaveBeenCalled();
  });
});

// // Snapshot testing

it("renders the button correctly when multiple props are set", () => {
  const { container } = render(
    <Button
      onClick={log}
      label="Snapshot button"
      kind="secondary"
      disabled
      pending
    />
  );
  expect(container).toMatchSnapshot();
});
