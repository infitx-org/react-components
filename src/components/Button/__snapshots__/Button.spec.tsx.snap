// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the button correctly when multiple props are set 1`] = `
<div>
  <button
    class="rc-button rc-button--secondary rc-button--large rc-button--disabled"
    disabled=""
    type="button"
  >
    <div
      class="rc-button__content"
    >
      <div
        class="rc-button__icon rc-button__icon--left"
      >
        <div
          class="rc-spinner rc-spinner--primary"
          style="width: 16px; height: 16px;"
        >
          <svg
            class="rc-spinner__component"
            height="16px"
            viewBox="0 0 16 16"
            width="16px"
          >
            <path
              class="rc-spinner__svg-path"
              d="M 5.811071082715721 14.014032773029815 A 6.4 6.4 0 0 0 14.4 8"
              stroke-width="1.6"
              style="stroke: inherit;"
            />
          </svg>
        </div>
      </div>
      Snapshot button
    </div>
  </button>
</div>
`;
